zerver/tests/test_auth_backends.py

```diff
@@ -1039,32 +1047,69 @@ def stage_two_of_registration(self, result: HttpResponse, realm: Realm, subdomai
                                  parsed_url.path)
         self.assertTrue(uri.startswith('http://zulip.testserver/accounts/login/subdomain/'))
 
-        result = self.client_get(result.url)
+        if not user_avatar_url:
+            result = self.client_get(result.url)
+        else:
+            data['user_avatar_url'] = user_avatar_url
+            data['next'] = ''
+            token = store_login_data(data)
+            url_path = reverse('zerver.views.auth.log_into_subdomain', args=[token])
+            result = self.client_get(url_path, subdomain=subdomain)
 
         self.assertEqual(result.status_code, 302)
         confirmation = Confirmation.objects.all().last()
         confirmation_key = confirmation.confirmation_key
         self.assertIn('do_confirm/' + confirmation_key, result.url)
+        if user_avatar_url:
+            # If there is an avatar_url, then there must be a user_avatar_url
+            # on the preregistration user.
+            prereg_user = confirmation.content_object
+            self.assertIsNotNone(prereg_user.user_avatar_url)
+
         result = self.client_get(result.url)
+
         self.assert_in_response('action="/accounts/register/"', result)
-        data = {"from_confirmation": "1",
-                "key": confirmation_key}
+        if user_avatar_url:
+            data = {"from_confirmation": "1",
+                    "user_avatar_url": user_avatar_url,
+                    "key": confirmation_key}
+        else:
+            data = {"from_confirmation": "1",
+                    "key": confirmation_key}
+
         result = self.client_post('/accounts/register/', data)
         if not skip_registration_form:
             self.assert_in_response("We just need you to do one last thing", result)
 
             # Verify that the user is asked for name but not password
             self.assert_not_in_success_response(['id_password'], result)
             self.assert_in_success_response(['id_full_name'], result)
+            if user_avatar_url:
+                # Verify if there is a checkbox for the user choose to use their
+                # github's avatar.
+                self.assert_in_success_response(['use_social_avatar'], result)
             # Verify the name field gets correctly pre-populated:
             self.assert_in_success_response([expected_final_name], result)
 
-            # Click confirm registration button.
-            result = self.client_post(
```

Comment:
We don't want to manually fiddle with the internal of the proces like this. What we want is to do the original client_post, adding the appropriate value of `user_social_avatar` to the dict.  And this call should happen in a `responses.RequestsMock` context, configured to mock `user_avatar_url` to return a realistic HTTP response with the image content. After that, we have blocks that verify correctness of the UserProfile that was created through this whole process. So we'll just want some asserts there to check that the user has (or hasn't) the correct avatar.

Reply:
@clarammdantas I think you missed this one comment? ^

Reply:
@clarammdantas So, instead of using this big if/else with a separate function, we could have something like this here:
```
with responses.RequestsMock(assert_all_requests_are_fired=True) as requests_mock:
    if user_avatar_url:
        requests_mock.add(user_avatar_url, ...)  # set up the appropriate HTTP response like github would give

    post_dict = {'full_name': expected_final_name,
                 'key': confirmation_key,
                 'terms': True})
    if user_avatar_url:
        post_dict['use_social_avatar'] = 'on' if use_social_avatar else ''

    result = self.client_post('/accounts/register/', post_dict)
```

Reply:
Oh, this also shows that we really need another argument to this stage_two_... function - in which we need to pass the image content that we want the avatar to have - that'll be much clearer then just implicitly assuming that it has to be `'zerver/tests/images/img.png'`.

Reply:
And then, it may make sense to make a simple `def verify_user_avatar(self, user_profile: UserProfile, expected_avatar_content_file: str) -> None:` helper that will assert the given `user_profile` has the avatar matching the content of the file - which you can call in this big verification block at the end of the function:
```
        if mobile_flow_otp:
            self.assertEqual(result.status_code, 302)
            redirect_url = result['Location']
            parsed_url = urllib.parse.urlparse(redirect_url)
            query_params = urllib.parse.parse_qs(parsed_url.query)
            self.assertEqual(parsed_url.scheme, 'zulip')
            self.assertEqual(query_params["realm"], ['http://zulip.testserver'])
            self.assertEqual(query_params["email"], [email])
            encrypted_api_key = query_params["otp_encrypted_api_key"][0]
            user_api_keys = get_all_api_keys(get_user_by_delivery_email(email, realm))
            self.assertIn(otp_decrypt_api_key(encrypted_api_key, mobile_flow_otp), user_api_keys)
            return
        elif desktop_flow_otp:
            self.verify_desktop_flow_end_page(result, email, desktop_flow_otp)
            # Now the desktop app is logged in, continue with the logged in check.
        else:
            self.assertEqual(result.status_code, 302)

        user_profile = get_user_by_delivery_email(email, realm)
        self.assert_logged_in_user_id(user_profile.id)
        self.assertEqual(user_profile.full_name, expected_final_name)

        self.assertFalse(user_profile.has_usable_password())
```
The `mobile_flow_otp` case has an early return, so you'd call the `self.verify_user_avatar` helper twice I guess, once in the `mobile_flow_otp` block, and once in the normal case, I guess right below here ro something:
```
        user_profile = get_user_by_delivery_email(email, realm)
        self.assert_logged_in_user_id(user_profile.id)
        self.assertEqual(user_profile.full_name, expected_final_name)
        self.verify_user_avatar(.....
```

Reply:
Hmm, I'm not sure that's what's supposed to be done. Because at this stage, I'm not handling a Github response anymore. I'm just handling with `prereg_user`, created at the beginning of `stage_two_of_registration` at:
```python
       self.assertTrue(uri.startswith('http://zulip.testserver/accounts/login/subdomain/'))

        result = self.client_get(result.url)
```
when it does that `self.client_get()`. Which will use the result data generated from calling `social_auth_test` function (who actually deals with the GitHub response.)

In `stage_two_of_registration`, `prereg_user` object will have all the information needed to decide whether or not the registration form will have a checkbox for the user to choose to use the avatar.

So, maybe a better approach will be add the response mocking a Github response, on the `social_auth_test`. And do almost like you did showing that mock, but mock the `upload_avatar_imagr_from_url` (because as the `user_avatar_url` doesn't really exist). So a draft of this last part would be:
```python
with mock.patch('zerver.lib.upload.upload_avatar_image_from_url') as mock_upload:
            mock_image_path = 'zerver/tests/images/img.png'
            mock_upload.return_value = user_avatar_content_hash(open(mock_image_path, 'rb').read())
            post_dict = {'full_name': expected_final_name,
                 'key': confirmation_key,
                 'terms': True})
             if user_avatar_url:
                 post_dict['user_avatar_url] = 'on' if use_social_avatar else ''

             result = self.client_post('/accounts/register/', post_dict)
```

Do the rest as you say with, creating the new function to verify the user avatar, and add the mock request mimicking the GitHub response on the `social_auth_test`. What do you think? Does that make sense?

Reply:
`prereg_user` has the URL, but doesn't know the content of the image - and we only fetch the image in that `/accounts/register/` codepath. So that's what we want to do - check that `accounts_register` will correctly fetch the image from the URL - and we check that by doing the POST inside the `RequestsMock` and then comparing the resulting user avatar with the expected content (from a file).
`social_auth_test` doesn't execute any codepath that fetches the URL content so it wouldn't make sense to do the mock there, unless I'm confused about something :thinking:

Reply:
But shouldn't the test of register a user be separated from the test that checks if we're fetching and uploading correctly the image associated with that URL? For me, I think at the `stage_two_of_registration` we should assume that the method that fetches and upload the image is correct. And I did a separate test to check if the `upload_avatar_image_from_url` is working fine.

Summarizing, I think that the responsibility of fetching and upload an image is from `upload_avatar_image_from_url` and should be a separate test. And in the registration, I should test only if the behavior of the user choosing or not to use the `social_avatar` is correct.

Reply:
Hmm, in general that could be a good line of reasoning I think. But as counterarguments:
1. In these social auth tests (and tbh this is our approach to tests generally) we test the flow end-to-end and (if you look at `social_auth_test`) the convention is that we mock the URLs from which data is fetched, rather than mocking the functions/methods that that do the fetching. It's nice to be consistent.
2. I think the test code will be a lot cleaner, doing a client_post  inside the simple `RequestsMock`, rather than how it is with the function mock? :thinking: 
3. I think the abstraction is nicer too with the URL mock - it's clearer 'ah, we're making the URL return the test image content!", rather than "we're mocking some function deep in the implementation details of user registration and creation". To trace what exactly is being mocked, one would have to dig through all the layers of functions being called, to find the specific function call in `create_user`. With URL mock one doesn't have to care at all how or where the fetching is implemented, it's just "the URL will return this content, we don't care about how exactly the request is made."

Reply:
aaaah ok, agreed. Now it makes sense to me. Thanks for the explanation! :)